## Course End Project
## Retail Analysis with Walmart Data
### By: Mike Dorado

### Problem Statement:

One of the leading retail stores in the US, Walmart, would like to predict the sales and demand accurately. There are certain events and holidays which impact sales on each day. There are sales data available for 45 stores of Walmart. The business is facing a challenge due to unforeseen demands and runs out of stock some times, due to the inappropriate machine learning algorithm. An ideal ML algorithm will predict demand accurately and ingest factors like economic conditions including CPI, Unemployment Index, etc.

Walmart runs several promotional markdown events throughout the year. These markdowns precede prominent holidays, the four largest of all, which are the Super Bowl, Labour Day, Thanksgiving, and Christmas. The weeks including these holidays are weighted five times higher in the evaluation than non-holiday weeks. Part of the challenge presented by this competition is modeling the effects of markdowns on these holiday weeks in the absence of complete/ideal historical data. Historical sales data for 45 Walmart stores located in different regions are available.


In [1]:
### Importing the packages
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
data=pd.read_csv('Walmart_Store_sales.csv')
data.head()

Store        Date  Weekly_Sales  Holiday_Flag  Temperature  Fuel_Price  \
0      1  05-02-2010    1643690.90             0        42.31       2.572   
1      1  12-02-2010    1641957.44             1        38.51       2.548   
2      1  19-02-2010    1611968.17             0        39.93       2.514   
3      1  26-02-2010    1409727.59             0        46.63       2.561   
4      1  05-03-2010    1554806.68             0        46.50       2.625   

          CPI  Unemployment  
0  211.096358         8.106  
1  211.242170         8.106  
2  211.289143         8.106  
3  211.319643         8.106  
4  211.350143         8.106

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6435 entries, 0 to 6434
Data columns (total 8 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Store         6435 non-null   int64  
 1   Date          6435 non-null   object 
 2   Weekly_Sales  6435 non-null   float64
 3   Holiday_Flag  6435 non-null   int64  
 4   Temperature   6435 non-null   float64
 5   Fuel_Price    6435 non-null   float64
 6   CPI           6435 non-null   float64
 7   Unemployment  6435 non-null   float64
dtypes: float64(5), int64(2), object(1)
memory usage: 402.3+ KB


In [4]:
data.isna().sum()

Store           0
Date            0
Weekly_Sales    0
Holiday_Flag    0
Temperature     0
Fuel_Price      0
CPI             0
Unemployment    0
dtype: int64

In [5]:
data.duplicated().sum()

0

In [6]:
### Convert data format to datetime
from datetime import datetime
data['Date']=pd.to_datetime(data['Date'],dayfirst=True)

In [7]:
data.dtypes

Store                    int64
Date            datetime64[ns]
Weekly_Sales           float64
Holiday_Flag             int64
Temperature            float64
Fuel_Price             float64
CPI                    float64
Unemployment           float64
dtype: object

In [8]:
total_sales=data.groupby('Store')['Weekly_Sales'].sum().round().sort_values(ascending=False)

In [9]:
total_sales.head(2)

Store
20    301397792.0
4     299543953.0
Name: Weekly_Sales, dtype: float64

#### Store ID # 20 - highest sales

In [10]:
pd.DataFrame(total_sales).tail(1)

Weekly_Sales
Store              
33       37160222.0

#### Store ID # 33 - lowest sales

#### Which store has the maximum standard diviation - i.e., the sales vary a lot. Also, find out the coefficient of mean to the standard diviation

In [11]:
data_std=data.groupby('Store')['Weekly_Sales'].std().round().sort_values(ascending=0)

In [12]:
pd.DataFrame(data_std).head(1)

Weekly_Sales
Store              
14         317570.0

#### Store # 14 has the highest variation in sales

In [13]:
store14=data[data.Store==14].Weekly_Sales

In [14]:
coefficient_of_variatition=store14.std()/store14.mean()*100
coefficient_of_variatition.round(2)

15.71

#### Coefficient of variatition for Store 14: 15.71%

#### Which store/s has good quarterly growth rate in Q3'2012

In [15]:
### Q2 Sales
q2_sales=data[(data['Date']>='2012-04-01')&(data['Date']<='2012-06-30')].groupby('Store')['Weekly_Sales'].sum().round()
q3_sales=data[(data['Date']>='2012-07-01')&(data['Date']<='2012-09-30')].groupby('Store')['Weekly_Sales'].sum().round()

In [16]:
pd.DataFrame({'Q2_Sales':q2_sales,'Q3_Sales':q3_sales,'Difference':(q3_sales-q2_sales),'Growth_Rate':(q3_sales-q2_sales)/q2_sales*100}).sort_values(by=['Growth_Rate'],ascending=0).head()

Q2_Sales    Q3_Sales  Difference  Growth_Rate
Store                                                 
7       7290859.0   8262787.0    971928.0    13.330775
16      6564336.0   7121542.0    557206.0     8.488383
35     10838313.0  11322421.0    484108.0     4.466636
26     13155336.0  13675692.0    520356.0     3.955475
39     20214128.0  20715116.0    500988.0     2.478405

#### Sales Impacts - which Holidays which have a higher sales than the mean sales in non-holiday season for all stores togather
#### Holidays: 
##### Super_Bowl: 12-Feb-10, 11-Feb-11, 10-Feb-12, 8-Feb-13
##### Labour_Day: 10-Sep-10, 9-Sep-11, 7-Sep-12, 6-Sep-13
##### Thanksgiving: 26-Nov-10, 25-Nov-11, 23-Nov-12, 29-Nov-13
##### Christmas: 31-Dec-10, 30-Dec-11, 28-Dec-12, 27-Dec-13

In [17]:
### Holiday Lists
Super_Bowl=['12-2-2010', '11-2-2011', '10-2-2012', '8-2-2013']
Labour_Day=['10-9-2010', '9-9-2011', '7-9-2012', '6-9-2013']
Thanksgiving=['26-11-2010', '25-11-2011', '23-11-2012', '29-11-2013']
Christmas=['31-12-2010', '30-12-2011', '28-12-2012', '27-12-2013']

In [20]:
import warnings
warnings.filterwarnings('ignore')

# Convert to Datetime
Super_Bowl=pd.to_datetime(Super_Bowl, dayfirst=True,format='mixed')
Labour_Day=pd.to_datetime(Labour_Day, dayfirst=True,format='mixed')
Thanksgiving=pd.to_datetime(Thanksgiving, dayfirst=True,format='mixed')
Christmas=pd.to_datetime(Christmas, dayfirst=True,format='mixed')


Super_Bowl_sales=round(data.loc[data.Date.isin(Super_Bowl)]['Weekly_Sales'].mean(),2)
Labour_Day_sales=round(data.loc[data.Date.isin(Labour_Day)]['Weekly_Sales'].mean(),2)
Thanksgiving_sales=round(data.loc[data.Date.isin(Thanksgiving)]['Weekly_Sales'].mean(),2)
Christmas_sales=round(data.loc[data.Date.isin(Christmas)]['Weekly_Sales'].mean(),2)

In [21]:
Super_Bowl_sales, Labour_Day_sales, Thanksgiving_sales, Christmas_sales


(1079127.99, 1042427.29, 1471273.43, 960833.11)